### PR TITLE
Share Query Cache feature

### PR DESCRIPTION
**Pull request by SergeiPavlov**
Setting `DomainConfiguration.ShareQueryCacheOverNodes = true` enables to share compiled queries between nodes

### CODE REVIEW

**Comment by alex-kulakov** [comment]
> I have little knowledge how to write multi-node DO test.
You can ask for help. I'll guide and help you, not a big deal.

>  But we are using this feature many months in production.
No offence but it'd rather choose proper test then "it works in our environment". I don't know your environment, I don't know how you use it. As I said I'll help you with tests.

**Comment by alex-kulakov** [comment]
I don't like upper-level types like ```TypeInfo``` and ```StorageNode``` got into SQL level. SQLDOM doesn't have to know about domains, sessions, typeinfos, storage nodes, etc. It is basically stand-alone, having just an SqlDriver, it should be allowed to extract schemas and catalogs, create and compile any SQL query supported by SQLDOM.

There is got to be a way to create an equivalent of the query created by DO by using SQLDOM and certain SqlDriver only, in this case it is impossible.

File: Orm/Xtensive.Orm/Sql/Compiler/SqlTranslator.cs [hunk 1]
```diff
@@ -1515,6 +1516,9 @@ public virtual void Translate(SqlCompilerContext context, object literalValue)
         case TimeSpan timeSpan:
           _ = output.Append(SqlHelper.TimeSpanToString(timeSpan, TimeSpanFormatString));
           break;
+        case TypeInfo typeInfo:
+          output.AppendPlaceholderWithId(typeInfo);
+          break;
```
Comment (alex-kulakov) [inline Orm/Xtensive.Orm/Sql/Compiler/SqlTranslator.cs:1521]: For example this is completely inappropriate.

File: Orm/Xtensive.Orm/Sql/Compiler/SqlCompiler.cs [hunk 2]
```diff
@@ -37,13 +38,13 @@ public class SqlCompiler : SqlDriverBound,
     /// <param name="unit">Instance to compile.</param>
     /// <param name="compilerConfiguration">Configuration for compiler.</param>
     /// <returns></returns>
-    public SqlCompilationResult Compile(ISqlCompileUnit unit, SqlCompilerConfiguration compilerConfiguration)
+    public SqlCompilationResult Compile(ISqlCompileUnit unit, SqlCompilerConfiguration compilerConfiguration, TypeIdRegistry typeIdRegistry = null)
```
Comment (alex-kulakov) [inline Orm/Xtensive.Orm/Sql/Compiler/SqlCompiler.cs:41]: This is also not ok, you have a configuration that wraps additional info for compilation. But no upper-level types should be used there.

**Comment by alex-kulakov** [comment]
I'm also thinking of having caches within nodes in cases when ```DomainConfiguration.ShareQueryCacheOverNodes = false``` but I'm weighting the outcome of this. Now number of items in query cache is multiplied by number of nodes, it should affect performance on thousands of nodes and thousands of queries (btw, how big is your cache size?).

Logically the idea of separated caches fits the concept attachable/detachable node - once node is disposed the resources it occupied should be released too.

**Comment by SergeiPavlov** [comment]
> There is got to be a way to create an equivalent of the query created by DO by using SQLDOM and certain SqlDriver only, in this case it is impossible.

I agree this is some kind of hack. But it just works, that is why I decided to use it.

**Comment by alex-kulakov** [comment]
Can you list the parts of query you replace with placeholders or something similar to make query to be cacheable?

**Comment by SergeiPavlov** [comment]
> Can you list the parts of query you replace with placeholders or something similar to make query to be cacheable?

* schema
* TypeId

**Comment by alex-kulakov** [comment]
Long time ago I did TypeId as parameter of query, but Alexander Ustinov said that there is something wrong with my implementation if applied to your case so i didn't merge it. Now I have a second thought that maybe it will somehow fit to this concept, ShareQueryCacheOverNodes  requires basically two features combined - TypeIdAsParameter and ShareSchemaOverNodes, doesn't it?

**Comment by SergeiPavlov** [comment]
> requires basically two features combined - TypeIdAsParameter and ShareSchemaOverNodes, doesn't it?

Looks like, yes
If you are talkimg about your year-old implementation of sharedCache-like feature,  I considered to use, but was not suitable for our case (does not remember why already)

**Comment by alex-kulakov** [comment]
I have improved TypeId As Parameter PR (#147), It is based on QueryParameterBinding now which is much better.

**Comment by alex-kulakov** [comment]
Additionally, I was inspired by your idea and changed schema actualization, here's the PR (#289). I think we could combine these two. PRs (#147 and #289) and just introduce condition like

```csharp
var domainConfig = domain.Configuration;
this.queryKey = domainConfig.ShareSchemaOverNodes && domainConfig.PreferTypeIdsAsQueryParameters
  ? queryKey
  : new Pair<object, string>(queryKey, session.StorageNodeId);
```  
in this PR.

**Comment by SergeiPavlov** [comment]
> just introduce condition like

done

**Comment by alex-kulakov** [comment]
The PR is ready - #290

**Pull request closed without merging**